File: mobile/lib/core/weddings_archive_index.dart
import 'package:hsdash_mobile/core/calendar_utils.dart';
import 'package:hsdash_mobile/models/shoot_calendar_entry.dart';
import 'package:hsdash_mobile/models/shoot_client_profile.dart';

/// One wedding (client) with one or more shoot events in a month.
class WeddingArchiveGroup {
  const WeddingArchiveGroup({
    required this.key,
    required this.displayName,
    required this.events,
  });

  final String key;
  final String displayName;
  final List<ShootCalendarEntry> events;
}

class WeddingsArchiveIndex {
  WeddingsArchiveIndex._(this._byYearMonth);

  final Map<int, Map<int, List<ShootCalendarEntry>>> _byYearMonth;

  factory WeddingsArchiveIndex.fromEntries(List<ShootCalendarEntry> entries) {
    final byYearMonth = <int, Map<int, List<ShootCalendarEntry>>>{};
    for (final e in entries) {
      final key = shootDayKey(e.day);
      final parts = key.split('-').map(int.parse).toList();
      final y = parts[0];
      final m = parts[1];
      byYearMonth.putIfAbsent(y, () => {});
      byYearMonth[y]!.putIfAbsent(m, () => []).add(e);
    }
    for (final months in byYearMonth.values) {
      for (final list in months.values) {
        list.sort((a, b) => shootDayKey(a.day).compareTo(shootDayKey(b.day)));
      }
    }
    return WeddingsArchiveIndex._(byYearMonth);
  }

  List<int> get years {
    final ys = _byYearMonth.keys.toList()..sort((a, b) => b.compareTo(a));
    return ys;
  }

  List<int> monthsForYear(int year) {
    final months = _byYearMonth[year]?.keys.toList() ?? [];
    months.sort((a, b) => a.compareTo(b));
    return months;
  }

  int weddingCountForYear(int year) {
    final keys = <String>{};
    for (final m in monthsForYear(year)) {
      keys.addAll(_weddingKeysForMonth(year, m));
    }
    return keys.length;
  }

  int weddingCountForMonth(int year, int month) => _weddingKeysForMonth(year, month).length;

  int eventCountForMonth(int year, int month) => _byYearMonth[year]?[month]?.length ?? 0;

  List<WeddingArchiveGroup> weddingsForMonth(int year, int month) {
    final entries = _byYearMonth[year]?[month] ?? [];
    final groups = <String, List<ShootCalendarEntry>>{};
    for (final e in entries) {
      final k = weddingKeyForEntry(e);
      groups.putIfAbsent(k, () => []).add(e);
    }
    final out = groups.entries
        .map(
          (g) => WeddingArchiveGroup(
            key: g.key,
            displayName: g.value.first.clientName,
            events: List<ShootCalendarEntry>.from(g.value)
              ..sort((a, b) => shootDayKey(a.day).compareTo(shootDayKey(b.day))),
          ),
        )
        .toList()
      ..sort((a, b) => a.displayName.compareTo(b.displayName));
    return out;
  }

  WeddingArchiveGroup? weddingGroup(int year, int month, String weddingKey) {
    for (final g in weddingsForMonth(year, month)) {
      if (g.key == weddingKey) return g;
    }
    return null;
  }

  Set<String> _weddingKeysForMonth(int year, int month) {
    final entries = _byYearMonth[year]?[month] ?? [];
    return entries.map(weddingKeyForEntry).toSet();
  }
}

/// Stable group id for the same couple/client across multiple shoot days.
String weddingKeyForEntry(ShootCalendarEntry entry) {
  final label = canonicalClientLabel(entry);
  if (label.isEmpty) return 'id:${entry.id}';
  return 'g:$label';
}

/// Display-style label used to match bride/groom rows with clientName-only rows.
String canonicalClientLabel(ShootCalendarEntry entry) {
  final bride = entry.brideName?.trim() ?? '';
  final groom = entry.groomName?.trim() ?? '';
  if (bride.isNotEmpty || groom.isNotEmpty) {
    return normalizeClientLabel(
      resolveShootClientName(isWedding: true, brideName: bride, groomName: groom, clientName: ''),
    );
  }
  return normalizeClientLabel(entry.clientName);
}

String normalizeClientLabel(String name) {
  var s = name.trim().toLowerCase();
  if (s.isEmpty) return '';
  s = s.replaceAll(RegExp(r'\s+'), ' ');
  s = s.replaceAll(RegExp(r'\s*&\s*'), ' and ');
  s = s.replaceAll(RegExp(r'[^a-z0-9 ]'), '');
  s = s.replaceAll(RegExp(r'\s+'), ' ').trim();
  return s;
}

/// All shoots for the same client/couple, sorted by day (earliest first).
List<ShootCalendarEntry> shootsForSameClient(
  List<ShootCalendarEntry> all,
  ShootCalendarEntry anchor,
) {
  final key = weddingKeyForEntry(anchor);
  return all
      .where((e) => weddingKeyForEntry(e) == key)
      .toList()
    ..sort((a, b) => shootDayKey(a.day).compareTo(shootDayKey(b.day)));
}

String monthLabel(int month) {
  const names = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];
  if (month < 1 || month > 12) return 'Month $month';
  return names[month - 1];
}

File: frontend/src/lib/weddingsArchiveIndex.ts
import { calendarDayKeyFromIso } from "@/lib/calendarUtils";
import { resolveShootClientName } from "@/types/shootClientProfile";
import type { ShootCalendarEntry } from "@/types/domain";

export function shootDayKey(day: string) {
  return calendarDayKeyFromIso(day);
}

export type WeddingArchiveGroup = {
  key: string;
  displayName: string;
  events: ShootCalendarEntry[];
};

export type WeddingsArchiveIndex = {
  years: number[];
  monthsForYear: (year: number) => number[];
  weddingCountForYear: (year: number) => number;
  weddingCountForMonth: (year: number, month: number) => number;
  eventCountForMonth: (year: number, month: number) => number;
  weddingsForMonth: (year: number, month: number) => WeddingArchiveGroup[];
  weddingGroup: (year: number, month: number, weddingKey: string) => WeddingArchiveGroup | null;
};

export function buildWeddingsArchiveIndex(entries: ShootCalendarEntry[]): WeddingsArchiveIndex {
  const byYearMonth = new Map<number, Map<number, ShootCalendarEntry[]>>();
  for (const e of entries) {
    const key = shootDayKey(e.day);
    const [y, m] = key.split("-").map(Number);
    if (!byYearMonth.has(y)) byYearMonth.set(y, new Map());
    const months = byYearMonth.get(y)!;
    if (!months.has(m)) months.set(m, []);
    months.get(m)!.push(e);
  }
  for (const months of byYearMonth.values()) {
    for (const list of months.values()) {
      list.sort((a, b) => shootDayKey(a.day).localeCompare(shootDayKey(b.day)));
    }
  }

  function monthsForYear(year: number) {
    const months = [...(byYearMonth.get(year)?.keys() ?? [])];
    months.sort((a, b) => a - b);
    return months;
  }

  function weddingKeysForMonth(year: number, month: number) {
    const monthEntries = byYearMonth.get(year)?.get(month) ?? [];
    return new Set(monthEntries.map(weddingKeyForEntry));
  }

  function weddingsForMonth(year: number, month: number) {
    const monthEntries = byYearMonth.get(year)?.get(month) ?? [];
    const groups = new Map<string, ShootCalendarEntry[]>();
    for (const e of monthEntries) {
      const k = weddingKeyForEntry(e);
      if (!groups.has(k)) groups.set(k, []);
      groups.get(k)!.push(e);
    }
    return [...groups.entries()]
      .map(([key, events]) => ({
        key,
        displayName: events[0]!.clientName,
        events: [...events].sort((a, b) => shootDayKey(a.day).localeCompare(shootDayKey(b.day))),
      }))
      .sort((a, b) => a.displayName.localeCompare(b.displayName));
  }

  return {
    years: [...byYearMonth.keys()].sort((a, b) => b - a),
    monthsForYear,
    weddingCountForYear(year) {
      const keys = new Set<string>();
      for (const m of monthsForYear(year)) {
        for (const k of weddingKeysForMonth(year, m)) keys.add(k);
      }
      return keys.size;
    },
    weddingCountForMonth(year, month) {
      return weddingKeysForMonth(year, month).size;
    },
    eventCountForMonth(year, month) {
      return byYearMonth.get(year)?.get(month)?.length ?? 0;
    },
    weddingsForMonth,
    weddingGroup(year, month, weddingKey) {
      return weddingsForMonth(year, month).find((g) => g.key === weddingKey) ?? null;
    },
  };
}

// Alias for call sites expecting static factory name
export const WeddingsArchiveIndex = { fromEntries: buildWeddingsArchiveIndex };

export function weddingKeyForEntry(entry: ShootCalendarEntry) {
  const label = canonicalClientLabel(entry);
  if (!label) return `id:${entry.id}`;
  return `g:${label}`;
}

export function canonicalClientLabel(entry: ShootCalendarEntry) {
  const bride = entry.brideName?.trim() ?? "";
  const groom = entry.groomName?.trim() ?? "";
  if (bride || groom) {
    return normalizeClientLabel(
      resolveShootClientName({ isWedding: true, brideName: bride, groomName: groom, clientName: "" }),
    );
  }
  return normalizeClientLabel(entry.clientName);
}

export function normalizeClientLabel(name: string) {
  let s = name.trim().toLowerCase();
  if (!s) return "";
  s = s.replace(/\s+/g, " ");
  s = s.replace(/\s*&\s*/g, " and ");
  s = s.replace(/[^a-z0-9 ]/g, "");
  s = s.replace(/\s+/g, " ").trim();
  return s;
}

export function shootsForSameClient(all: ShootCalendarEntry[], anchor: ShootCalendarEntry) {
  const key = weddingKeyForEntry(anchor);
  return all
    .filter((e) => weddingKeyForEntry(e) === key)
    .sort((a, b) => shootDayKey(a.day).localeCompare(shootDayKey(b.day)));
}

const MONTH_NAMES = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

export function monthLabel(month: number) {
  if (month < 1 || month > 12) return `Month ${month}`;
  return MONTH_NAMES[month - 1]!;
}
